Finish Cereal serialization

dkeeney opened the pull request:
This PR removes the old serialization and switches to Cereal serialization.

dkeeney commented:
This is in progress.  It still has a few bugs to track down but I am going away for a long weekend and wanted to get this saved.  Will be back late Monday.

dkeeney commented:
I spent a little more time on the bug that is crashing the ScalarEncoderTest after I switched to Cereal.  This is pointing out another restriction with Cereal.   

The ScalarEncoderTest.Serialization test iterates through a set of scenarios, saving each sequentially into a single stream.  Then it iterates through each scenario again and loads from the one stream.  With normal stream serialization this works fine because the parsing stops when the current request is satisfied.

However, with Cereal, on a load it parses the entire stream first and then restores the requested arguments.  What is happening is that on the load of the first scenario it parses everything but only uses the part that it needs.  The next scenario then tries to perform a load and tries to parse the stream but it is already at the end of the stream and so it crashes....or gives a read error.

I will change the test so that it uses a separate stream for each scenario, but I wanted everyone to know about this restriction.  I will also add something to Serializable.hpp to describe this problem.

dkeeney commented:
Please review when you have some time. I know this contains a lot of files to look through but it is just the cleanup needed to switch from the binary stream serialization to Cereal.
- A few of them I used the original save( ) function to create a human readable operator<<.
- The Dimension class continued to conflict with Cereal so I switched from a subclass of std::vector to a wrapper for std::vector.  This seems to work much better.
- The wrapper class BundleIO is no longer needed so BundleIO.hpp was removed.
- Otherwise it was only the things that had been marked for remove when the switch occurred.

After this is merged in we can start the next project which is to replace `nupic` with `htm`.

ctrl-z-9000-times commented on src/nupic/types/Sdr.hpp:
```diff
@@ -582,32 +584,55 @@ class SparseDistributedRepresentation : public Serializable
+        for(UInt dim : sdr.dimensions_)
+            sdr.size_ *= dim;
+        // Initialize sparse tuple.
+        sdr.coordinates_.assign( sdr.dimensions_.size(), {} );
```
Nitpick: call the initialize method?

dkeeney replied:
> Nitpick: call the initialize method?

Hmmm, you mean reshape( ) maybe?

This is in the `operator>>` function which parses the output generated by `operator<<`.
There is a minor problem...I have one a lambda function that parses the stream into a `vector<UInt>`.  That works for the dimensions.  However, for the sparse data, the type is `vector<ElemSparse>` (which happens to resolve to UInt).  If I used this same lambda function for both it will compile but if ElemSparse should change then we are screwed.  We don't have templated lambda functions until C++20 so I am forced to make the lambda function into a real private function.  
Do you have any other ideas?

ctrl-z-9000-times replied:
`ElemSparse` is going to be a UInt32 until someone wants to make an HTM with more than 4 billion cells.  I don't think this is going to be a concern for a while...

ctrl-z-9000-times replied:
On second thought, I think what you have here now is fine as is.

Currently `operator>>`  constructs the SDR (assigns new dimensions).
If `operator>>` calls resize then the SDR must already be initialized with the correct size (or it will throw exception).

IMO the best would be for `operator>>` to check if the SDR is initialized (has dimensions).  If the SDR is initialized then it should check dimensions or reshape it.  If the SDR is not initialized then it should just assign the dimensions.  This how SDR assigment `operator=` currently works.

ctrl-z-9000-times commented on src/test/unit/algorithms/SpatialPoolerTest.cpp:
```diff
@@ -1865,11 +1865,13 @@ TEST(SpatialPoolerTest, testSaveLoad) {
-  outfile.open(filename);
+  outfile.open(filename, ifstream::binary);
+	outfile.precision(std::numeric_limits<double>::digits10 + 1);
+	outfile.precision(std::numeric_limits<float>::digits10 + 1);
```
Should this precision be set by the serialize method?  Is this something that every user is going to need?

dkeeney replied:
> Should this precision be set by the serialize method?

This is what the saveToFile( ) serialization sets.  I guess in this test, I did not need to set it...it would default to whatever the max resolution is.  I will remove those so it will test that it does not need to be set.  The stream does need to be initialized as ifstream::binary however because it is taking the default serialization format which is BINARY.

ctrl-z-9000-times replied:
FYI This issue is also on lines:

https://github.com/htm-community/nupic.cpp/blob/418f51cf49ba928ec8f2487773f6f1ef8c13735d/src/test/unit/algorithms/SpatialPoolerTest.cpp#L1904-L1905

https://github.com/htm-community/nupic.cpp/blob/418f51cf49ba928ec8f2487773f6f1ef8c13735d/src/test/unit/algorithms/SpatialPoolerTest.cpp#L1935-L1936

https://github.com/htm-community/nupic.cpp/blob/418f51cf49ba928ec8f2487773f6f1ef8c13735d/src/test/unit/algorithms/SpatialPoolerTest.cpp#L1996-L1997

dkeeney replied:
> os.precision(std::numeric_limits<double>::digits10 + 1); 
os.precision(std::numeric_limits<float>::digits10 + 1); 

I removed the ones at line 1868 to confirm that they are not required.  I left in the other three places to the test can confirm that they can be specified by the user.

ctrl-z-9000-times replied:
Makes sense.

ctrl-z-9000-times commented on src/test/unit/algorithms/SpatialPoolerTest.cpp:
```diff
@@ -2080,15 +2088,24 @@ TEST(SpatialPoolerTest, testConstructorVsInitialize) {
   SDR gold_sdr;
-  gold_sdr.load( gold_stream );
+  gold_sdr.load( gold_stream, JSON );
+  EXPECT_EQ(silver_sdr, gold_sdr);
```
Can you remove two of these SDRs and keep just one?

dkeeney replied:
> Can you remove two of these SDRs and keep just one?

I was trying to test two different ways to load an SDR using deserialization.
- silver is loaded directly using initialization and that becomes the standard.
- gold is being loaded via Cereal from a manually written JSON string.
- copper is being loaded via operator>>.  The operator>> should be able to read the output from the operator<< output that you preferred.   

However,  I could not get the operator>> test to work right away.  It was kind of outside the scope of this PR and I did not want to spend a lot of time on it so I commented it out with a TODO.  I will spend a little more time on that.  I think that would be nice-to-have.

ctrl-z-9000-times replied:
Yes, it would be nice to have, but not critical.

An optional nitpick is to move this into the SDR unit tests.  My concern is that if/when the Spatial Pooler code changes (and we want to update the gold string) then we will have more strings to update.

dkeeney replied:
>  My concern is that if/when the Spatial Pooler code changes 

Well, yes that is a good point.  That is true will any test data put into any test.  

Ok, I will just remove the operator>> function and if it turns out to be needed it can be added with a PR that addresses this issue.  It really is sort of out of scope for this PR anyway.
There was not a strong reason to be making that change.

dkeeney commented on src/nupic/algorithms/SpatialPooler.cpp:
```diff
@@ -1144,40 +990,62 @@ void SpatialPooler::printParameters() const {
+  out << "minOverlapDutyCycles: ";
+  sp.printState(sp.minOverlapDutyCycles_, out);
+  out << "tieBreaker: ";
+  sp.printState(sp.tieBreaker_, out);
```
> I would really like to see some analysis in the print methods. 

I agree, that would be much more useful.  I was using the operator>> to create a string that can be compared with operator== which is used by the unit tests to see if we get the same object back after a deserialization.  I think you are right, this is not the right way to do this.  I will rewrite the operator== comparing everything...probably needs the tieBreakers too.

I do not know enough about the class to print any analysis and other things that might be important info.  So I will defer to your version of operator<< and we can
fix the extra things you suggested in another PR.

ctrl-z-9000-times replied:
I assumed that the == method would use the new serialization to compare.   Either way its up to you how you want to make == work as well as the unit tests.  

Personally, I would consider changing to unit tests to not use == and then dropping the method.  Instead, run a bunch of random data through the SP and check results.  But its up to you.

dkeeney replied:
> Personally, I would consider changing to unit tests to not use ==

I think a hard coded compare routine is the best.  Running data through SP might be 99.9999% sure but comparing all restored data directly would be 100% sure that the two were exactly the same after restore.  Besides, running SP would be MUCH slower.

I must admit that running SP many times would make sure that all of the RIGHT data had been saved and restored.  But that would show up with some of the other tests.

ctrl-z-9000-times replied:
Ok, that all makes sense.

dkeeney commented on src/nupic/algorithms/TemporalMemory.cpp:
```diff
@@ -1026,22 +822,84 @@ bool TemporalMemory::operator==(const TemporalMemory &other) const {
       << "permanenceIncrement       = " << getPermanenceIncrement() << std::endl
       << "permanenceDecrement       = " << getPermanenceDecrement() << std::endl
       << "predictedSegmentDecrement = " << getPredictedSegmentDecrement()
+      << "anomaly                   = " << anomaly_ << std::endl
```
> Anomaly is not a parameter.

I think what should be here in TemporalMemory is some analysis like mentioned for SpatialPooler. I was just converting the old save( ) function to operator<<.  Again, I am not qualified to know what should really be here so I will have to defer to another PR.  It is kind of out-of-scope for this PR anyway.

dkeeney commented on src/nupic/engine/Spec.cpp:
```diff
@@ -103,6 +103,16 @@ bool InputSpec::operator==(const InputSpec &o) const {
          count == o.count && description == o.description;
 }
+std::ostream& operator<< (std::ostream& out, const InputSpec& self) {
+   out << "     description: " << self.description << "\n"
```
> nitpick: use std::endl instead of \n for portability.

This is portable.  On windows, a '\n' is replaced with '\r\n' if written to a file.  The only difference between '\n' and std::endl is that the std::endl will also flush...sometimes.

dkeeney commented on src/nupic/encoders/ScalarEncoder.cpp:
```diff
@@ -195,45 +195,20 @@ void ScalarEncoder::encode(Real64 input, SDR &output)
+  out << "minimum: " << self.parameters.minimum    << ", ";
+  out << "maximum: " << self.parameters.maximum    << ", ";
+  out << "clipInput: " << self.parameters.clipInput  << ", ";
+  out << "periodic: " << self.parameters.periodic   << ", ";
```
> ScalarEncoder.cpp clean up the display

Done.

ctrl-z-9000-times replied:
I'm not seeing this change on my end...  Did you remember to add this file?

dkeeney replied:
hmmm, I did seem to have lost that change.  I will re-do it.

dkeeney commented on src/nupic/encoders/RandomDistributedScalarEncoder.cpp:
```diff
@@ -127,37 +127,13 @@ void RandomDistributedScalarEncoder::encode(Real64 input, sdr::SDR &output)
+  out << "size: " << self.parameters.size << ", ";
+  out << "activeBits: " << self.parameters.activeBits << ", ";
+  out << "resolution: " << self.parameters.resolution << ", ";
+  out << "category: " << self.parameters.category << ", ";
```
> RandomDistributedScalarEncoder.cpp ... cleanup the display

Done

dkeeney commented on src/nupic/algorithms/TemporalMemory.cpp:
```diff
@@ -1026,22 +822,84 @@ bool TemporalMemory::operator==(const TemporalMemory &other) const {
+  }
+  out << "]" << endl;
+
+  out << "activeCells: [ ";
```
> TemporalMemory.cpp Umm... This isn't what I envisioned for the print methods.

Yes, the old save( ) function really does not work very well as a human readable display even if it is text.  I am deleting it and defer to another PR to do it right.

dkeeney commented:
I think I have addressed all of the review points...good review by the way.
- removed operator>> from SDR.  Thought that would be an easy thing to add but caused more problems than it was worth and not really in scope.
- My poor attempts at operator<< for SpatialPooler and TemporalMemory were reverted and left for a later PR where they could be done correctly.
- hard coded the operator== for SpatialPooler object rather than comparing serialized strings.

ctrl-z-9000-times commented:
> removed operator>> from SDR. Thought that would be an easy thing to add but caused more problems than it was worth and not really in scope.

Thanks for doing this!  I copied your code into a private notebook, I will submit it in a separate PR when I get the chance.

ctrl-z-9000-times commented on src/nupic/algorithms/TemporalMemory.cpp:
```diff
@@ -846,74 +846,11 @@ void TemporalMemory::printParameters(std::ostream& out) const {
       << std::endl
       << "maxSegmentsPerCell        = " << getMaxSegmentsPerCell() << std::endl
       << "maxSynapsesPerSegment     = " << getMaxSynapsesPerSegment()
-      << "extra                     = " << extra_ << std::endl
```
Technically this is a parameter, but it's not critical.

ctrl-z-9000-times replied:
Please ignore this comment.

ctrl-z-9000-times commented on src/nupic/algorithms/TemporalMemory.cpp:
```diff
@@ -846,74 +846,11 @@ void TemporalMemory::printParameters(std::ostream& out) const {
       << "connectedPermanence       = " << getConnectedPermanence() << std::endl
       << "minThreshold              = " << getMinThreshold() << std::endl
       << "maxNewSynapseCount        = " << getMaxNewSynapseCount() << std::endl
-      << "checkInputs               = " << checkInputs_ << std::endl
```
checkInputs is also technically a parameter, but it's not really useful to know in this table.

checkInputs & extra probably should not be in this table of parameters since they don't really effect the functionality of the TM.  I think that your changes (removals) here should stand.

dkeeney commented:
So, do I need to do something more to get this PR approved?

dkeeney commented:
I re-did the changes to ScalarEncoder....added some \n's to the operator<< and made it a little more readable.

This means I will need another approval...sorry.

ctrl-z-9000-times commented:
The test that failed is known to be flaky...  I restarted the CI.

dkeeney commented:
Thank you.
